### PR TITLE
Grammar for Canvas API.

### PR DESCRIPTION
**Pull request by symeonp**
Adding initial grammar for Canvas API.

### CODE REVIEW

**Comment by symeonp** [comment]
I signed it!

**Comment by symeonp** [comment]
Hey Ivan, that's my very first attempt to create an initial grammar for the Canvas API!
I'll try to add more functions along the way!
Comments, ideas, domatos 🍅 are welcome :o)

File: canvas/canvas.txt [hunk 1]
```diff
@@ -0,0 +1,384 @@
...
+<largeint> = 2147483647
+<largeint> = 2147483648
+<largeint> = 4294967295
+<largeint> = 4294967296
+
+<fuzzstring> = <fuzzstringpart>
+<fuzzstring> = <fuzzstringpart> + <fuzzstringpart>
+<fuzzstring> = <fuzzstringpart> + <fuzzstringpart> + <fuzzstringpart>
+<fuzzstringpart> = "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA"
+<repeatcount> = 17
```
Comment (ifratric) [inline canvas/canvas.txt]: `<repeatcount>` is never used.

File: canvas/canvas.txt [hunk 2]
```diff
@@ -0,0 +1,384 @@
...
+<repeatcount> = 65
+<repeatcount> = 257
+<repeatcount> = 1025
+<repeatcount> = 4097
+<repeatcount> = 65537
+<repeatstr> = String.fromCharCode(<int min=0 max=127>)
```
Comment (ifratric) [inline canvas/canvas.txt]: Releated to the above comment, you probably want something along the lines of 
`<repeatstr> = String.fromCharCode(<int min=0 max=127>).repeat(<repeatcount>)`

File: canvas/canvas.txt [hunk 3]
```diff
@@ -0,0 +1,384 @@
...
+!varformat fuzzvar%05d
+!begin lines
+!include common.txt
+
+# Image
+<new img> = new Image();
+<new img> = new Image(<fuzzint>, <fuzzint>);
+<new img> = new Image(<largeint>, <largeint>);
+<new img> = new Image(<interestingint>, <interestingint>);
+<new img> = new Image(<this>, <interestingint>);
```
Comment (ifratric) [inline canvas/canvas.txt]: Why the `this` object here? Wouldn't this be equivalent of setting the width to 0

File: canvas/canvas.txt [hunk 4]
```diff
@@ -0,0 +1,384 @@
...
+<globalCompositeOperation> = "luminosity"
+
+<this> = this
+
+<root root=true> = <lines count=50>
+
+#!lineguard try { <line>; } catch(e) { console.log(e.message) }
+!varformat fuzzvar%05d
+!begin lines
+!include common.txt
```
Comment (ifratric) [inline canvas/canvas.txt]: You include common.txt here, but above you already declared some symbols with the same names as in common.txt, for example `<color>` and `<fuzzint>`. The effect is going to be that these are going to be mixed together, which I'm not sure is what you intended. I suggest either relying on values from common.txt or renaming the new symbols you declare here to avoid the mixup.

File: canvas/canvas.txt [hunk 5]
```diff
@@ -0,0 +1,384 @@
...
+<new img> = new Image();
+<new img> = new Image(<fuzzint>, <fuzzint>);
+<new img> = new Image(<largeint>, <largeint>);
+<new img> = new Image(<interestingint>, <interestingint>);
+<new img> = new Image(<this>, <interestingint>);
+<new img> = new Image(<interestingint>, <this>);
+<img>.src = "data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///ywAAAAAAQABAAACAUwAOw==";
+
+# Uint8ClampedArray
+<new uintc8> = new Uint8ClampedArray(<fuzzint>);
```
Comment (ifratric) [inline canvas/canvas.txt]: Instead of repeating the lines like this, you can declare a single integer type, e.g.

```
<canvasint> = <interestingint>
<canvasint> = <fuzzint>
<canvasint> = <largeint>
```

and then just write

`<new uintc8> = new Uint8ClampedArray(<canvasint>);`

Which avoids a lot of line duplication here and in other places.

File: canvas/generator.py [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+import re
+import random
+import sys
+
+parent_dir = os.path.abspath(os.path.join(os.path.dirname(__file__), os.pardir))
+sys.path.append(parent_dir)
+from grammar import Grammar
+
+_N_MAIN_LINES = 1000
+_N_EVENTHANDLER_LINES = 500
```
Comment (ifratric) [inline canvas/generator.py]: This is never used

File: canvas/generate_one_case.py [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+import sys
+parent_dir = os.path.abspath(os.path.join(os.path.dirname(__file__), os.pardir))
+sys.path.append(parent_dir)
+from grammar import Grammar
+
+my_grammar = Grammar()
+my_grammar.parse_from_file('canvas.txt')
+result_string = my_grammar.generate_root()
+
+print (result_string)
```
Comment (ifratric) [inline canvas/generate_one_case.py]: Is this file really needed? Doesn't generator.py already have all that is done here?

File: canvas/template.html [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+
+function GetVariable(fuzzervars, var_type) { if(fuzzervars[var_type]) { return fuzzervars[var_type]; } else { return null; }}
+
+function SetVariable(fuzzervars, var_name, var_type) { fuzzervars[var_type] = var_name; }
+
+function draw() {
```
Comment (ifratric) [inline canvas/template.html]: Where is `draw()` called from?

File: canvas/template.html [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+    canvas.height = 200;
+    canvas.style.position = "absolute";
+    canvas.style.border = "1px solid";
+
+    var body = document.getElementsByTagName("body")[0];
+    body.appendChild(canvas);
+
+    var ctx = canvas.getContext('2d');
+
+    try { <canvasfuzz>; } catch(e) { console.log(e.message) }
```
Comment (ifratric) [inline canvas/template.html]: why have try/catch around `<canvasfuzz>` when each line of `canvasfuzz` is also going to be surrounded by try/catch?

File: canvas/canvas.txt (hunk 2 above)
Comment (symeonp) [reply canvas/canvas.txt]: Indeed you are right, I updated it!

File: canvas/canvas.txt (hunk 3 above)
Comment (symeonp) [reply canvas/canvas.txt]: Right, I've removed that as well!

File: canvas/canvas.txt (hunk 4 above)
Comment (symeonp) [reply canvas/canvas.txt]: Indeed, I slightly messed this one up, updated it and using the values from common.txt

File: canvas/canvas.txt (hunk 5 above)
Comment (symeonp) [reply canvas/canvas.txt]: You are *so* right, I added so much redundant code and I ended up getting bored (programming is tough!) - I've cleaned up this mess!

File: canvas/generator.py (hunk 6 above)
Comment (symeonp) [reply canvas/generator.py]: Now added these lines that it does:

```
    handlers = False
    while '<jsfuzzer>' in result:
        numlines = _N_MAIN_LINES
        if handlers:
            numlines = _N_EVENTHANDLER_LINES
        else:
            handlers = True
```

File: canvas/generate_one_case.py (hunk 7 above)
Comment (symeonp) [reply canvas/generate_one_case.py]: Well, that was quick silly test, instead of running every time generator.py with arguments and writing to files and stuff, I've used it to quickly check the output and modify the grammar based on what I needed!

File: canvas/template.html (hunk 8 above)
Comment (symeonp) [reply canvas/template.html]: That is used from my harness (I will release once I've manage to get crashes actually!)
- updated the file as well.

File: canvas/template.html (hunk 9 above)
Comment (symeonp) [reply canvas/template.html]: If you notice from my grammar:
#!lineguard try { <line>; } catch(e) { console.log(e.message) }
this line is actually commented out, hence it will never run.
I removed it :)

**Comment by symeonp** [comment]
I've done all the changes here:
https://github.com/symeonp/domato/commit/f33255b40a6de37b9013ea18bb4b5470a3d0aa56
If you are happy, please let me know, and we can close this PR!
Thank you so much for your comments, hopefully it's a bit cleaner now :)

**Comment by ifratric** [comment]
Yep, much cleaner :-)

But, why did you remove the !lineguard in the latest update: The effect of this is that as soon as any of the lines throws an exception (due to e.g. invalid param), none of the other lines are going to execute. I suggest returning it and removing try/catch around <canvasfuzz> instead.

Also, consider adding <script> tags in template so that it can be run in the browser simply by opening the output file.

**Comment by symeonp** [comment]
Done! Brilliant,thank you!
